### PR TITLE
Feature/thumbnail loading placeholder

### LINKED ISSUE #161: Loading Icon for rendered Zarr images
Since this can take some time w/ the zarr network calls, we might consider adding a loading icon specific to the thumbnail, especially since otherwise it can display the wrong thumbnail for a while

_Originally posted by @aswallace in https://github.com/AllenInstitute/aics-fms-file-explorer-app/pull/143#discussion_r1664915655_

### PR DESCRIPTION
**Pull request by BrianWhitneyAI**
## Description 

The purpose of this PR is to resolve #161 and add a loading state to thumbnails that we are rendering. In the course of this PR the following logic was added:

- Loading state for thumbnail processes
- Try logic / failed state for zarr thumbnails
- Retry logic for Zarr thumbnails (3) + (Timeout)

#### Small icon loading 
Screenshot 2024-10-11 at 12 11 01 PM
#### Failed Zarr (top left corner)
Screenshot 2024-10-11 at 12 10 43 PM

### CODE REVIEW

File: packages/core/components/FileThumbnail/index.tsx [hunk 1]
```diff
@@ -11,12 +12,29 @@ interface Props {
...
+    // If loading, render the spinner instead of the image
+    if (props.loading) {
+        return (
+            <div
+                className={classNames(
+                    props.className,
+                    styles.fileThumbnail,
+                    styles.thumbnailPlaceholder
+                )}
+                style={{ height: props.height, width: props.width }}
```
Comment (SeanLeRoy) [inline packages/core/components/FileThumbnail/index.tsx:31]: Should this default to being a square if one of the fields is present but not the other? Like so:
```
style={{ height: props.height || props.width, width: props.width || props.height }}
```

**Review by SeanLeRoy — APPROVED** [review]

**Comment by BrianWhitneyAI** [comment]
Currently has a visual bug where the loading box / spinner is misaligned. I tried recentering but somethings weird. 
Screenshot 2024-10-15 at 12 35 45 PM

File: packages/core/components/FileThumbnail/FileThumbnail.module.css [hunk 2]
```diff
@@ -8,3 +8,11 @@
 .no-thumbnail {
     fill: var(--border-color);
 }
+
+.thumbnailPlaceholder {
+    display: flex;
+    justify-content: center;
+    align-items: center;
+    background-color: var(--border-color);
+    border: 5px var(--primary-background-color);
```
Comment (aswallace) [inline packages/core/components/FileThumbnail/FileThumbnail.module.css:17]: Adding `margin: auto;` here should take care of the centering issue

File: packages/core/entity/FileDetail/RenderZarrThumbnailURL.ts [hunk 3]
```diff
@@ -38,89 +38,131 @@ function transformAxes(originalArray: AxisData[]): TransformedAxes[] {
...
+
+/**
+ * Main function to attempt to render a usable thumbnail using the lowest
+ * resolution present in a zarr image's metadata.
+ */
+export async function renderZarrThumbnailURL(zarrUrl: string): Promise<string | undefined> {
+    try {
+        return await retryWithTimeout(
+            async () => {
+                console.log("Starting Zarr thumbnail rendering for:", zarrUrl);
```
Comment (aswallace) [inline packages/core/entity/FileDetail/RenderZarrThumbnailURL.ts:79]: Do we still need the log here?

**Review by aswallace — APPROVED** [review]

File: packages/core/components/FileThumbnail/index.tsx (hunk 1 above)
Comment (BrianWhitneyAI) [reply packages/core/components/FileThumbnail/index.tsx]: resolved in [478fbd2](https://github.com/AllenInstitute/biofile-finder/pull/268/commits/478fbd2d80b96884e30648830079e4285557fbf1)

File: packages/core/entity/FileDetail/RenderZarrThumbnailURL.ts (hunk 3 above)
Comment (BrianWhitneyAI) [reply packages/core/entity/FileDetail/RenderZarrThumbnailURL.ts]: resolved in [dcce3a5](https://github.com/AllenInstitute/biofile-finder/pull/268/commits/dcce3a53988fff6be5dbfdfd0dd0d19fcbda3357)

File: packages/core/components/FileThumbnail/FileThumbnail.module.css (hunk 2 above)
Comment (BrianWhitneyAI) [reply packages/core/components/FileThumbnail/FileThumbnail.module.css:17]: resolved in [dcce3a5](https://github.com/AllenInstitute/biofile-finder/pull/268/commits/dcce3a53988fff6be5dbfdfd0dd0d19fcbda3357)

File: packages/core/entity/FileDetail/RenderZarrThumbnailURL.ts [hunk 4]
```diff
@@ -38,89 +38,129 @@ function transformAxes(originalArray: AxisData[]): TransformedAxes[] {
...
 }
+/**
+ * Helper function to handle retry logic with timeout for async operations.
+ * It retries the operation up to the specified number of times and aborts if it takes too long.
+ */
+async function retryWithTimeout<T>(fn: () => Promise<T>, retries = 3, timeout = 5000): Promise<T> {
+    let attempt = 0;
+    while (attempt < retries) {
+        try {
+            return await withTimeout(fn(), timeout);
```
Comment (SeanLeRoy) [inline packages/core/entity/FileDetail/RenderZarrThumbnailURL.ts:49]: What is the purpose of the timeout here. Does it help the callstack avoid getting overloaded somehow?

Comment (BrianWhitneyAI) [reply packages/core/entity/FileDetail/RenderZarrThumbnailURL.ts:49]: From what i can tell some async requests in the zarr render dont register as  failures, This just says if its taking too long count it as a failure.

**Pull request merged**